Add "Quick Start" tutorial

Sbozzolo opened the pull request:
This PR adds a "Quick Start" tutorial that walks new users through their first simulation by providing a different view on the "two sphere" example. 

The aim of the tutorial is to expose new users to the basic concepts involved with installing palace and running and analyzing simulations. `quick.md` leverages Documenter.jl's ability to execute code upon doc generation, so that many sections in the tutorial are populated automatically. This has two consequences:
- the "quick start" experience itself is somewhat tested,
- many updates to the two sphere examples or to palace are automatically reflected to the rendered tutorial.
For this to happen, I updated the docs.yml to install palace with spack. In order to avoid a 30+ minute build, the workflow has to be configured to ensure that it hits the GHCR cache.

The tutorial does not cover meshing. Instead, this PR also adds comments to the `examples/spheres/mesh.jl` file, with the hope to be more accessible to users with no prior gmsh experience.

Read it here:
https://awslabs.github.io/palace/previews/PR415/quick/

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -15,6 +26,66 @@ jobs:
+                root: /opt/spack
+                padded_length: False
+            concretizer:
+              reuse: false
```
```suggestion
              reuse: true
```

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -15,6 +26,66 @@ jobs:
+            packages:
+              petsc:
+                require: ~hdf5
+              rocblas:
+                require: ~tensile
+            repos:
+            - spack/local
```
```suggestion
```

As discussed, not needed for this workflow

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -15,6 +26,66 @@ jobs:
+                require: ~tensile
+            repos:
+            - spack/local
+            mirrors:
```
These are evaluated in-order FYI

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -15,6 +26,66 @@ jobs:
+
+      - name: Install Palace
+        run: spack  -e . install palace
+
```
For reference

```
      # Push built binaries, even if the build fails
      # NOTE: This might fail as external fork PRs can't push to GHCR.
      - name: Push to GHCR cache
        if: |
          !cancelled() &&
          (github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == github.repository)
        run: spack -e . buildcache push --force --with-build-dependencies --unsigned --update-index local-buildcache
```

cameronrutherford reviewed (commented):
Small comments per offline discussion

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -4,9 +4,20 @@ on:
+
 jobs:
   build-docs:
     permissions:
```
```suggestion
    permissions:
      packages: write
```

This is the only other thing that I can think of from the slack thread... It seems like we are doing everything right.

The other alternatives I can think of are:
- Just to abandon GHCR for this and use an S3 bucket
- Use different GHCR caches for each workflow

I maintain that this should just work....

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -7,18 +7,87 @@ on:
+        with:
+          ref: develop
+          buildcache: true
+          color: true
+          # This seems unnecessary, but otherwise we clone into ./spack
+          # This then would override the spack local repo dir, and break CI
+          path: spack-path
```
With #427, we no longer need this with our internal spack packages repo changing it's pathname.

In addition to that, this workflow doesn't use that in the first place, so we don't need this.

```suggestion
          color: true
```

(I don't think GitHub is liking this sort of "only deleting lines" inline suggestion, so this might have to be manual)

Sbozzolo commented on docs/src/quick.md:
````diff
@@ -0,0 +1,542 @@
+palace -np 1 spheres.json
+```
+
+`-np 1` instruct *Palace* to run on with single MPI process.
````
Yes, typo!

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,542 @@
+
+The `Problem` section identifies the problem type and the output directory. In
+this case, we choose `Electrostatic`. This means that *Palace* solves Poisson's
+equation for electric potential sequentially activating all the `Terminals` on
+the mesh. All simulation types in *Palace* have some form of iteration (over
```
, whilst setting the non-activated terminals to ground.

that the other terminals are held at ground whilst this is occurring is important.

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,542 @@
+The `Domains` section also includes a `Postprocessing` subsection for
+calculating specific quantities. In this example, we add:
+
+  - `Energy` activates integrated energy calculations in the 3D domain
```
`Energy` which activates.

The grammar is a bit weird without the connective.

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,542 @@
+calculating specific quantities. In this example, we add:
+
+  - `Energy` activates integrated energy calculations in the 3D domain
+  - `Probe` requests field values at specific coordinates defined by a `Center`
```
`Probe` which requests

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,542 @@
+    `Attributes` identify mesh regions and come from the mesh file. In our example,
+    attributes 1-4 identify the vacuum region, outer boundary, and two spheres.
+    
+    `Index` is used only for postprocessing and defines a notation used in the output CSV files. It has no relation to mesh attributes and can
+    be any positive integer.
```
Maybe note that there can be multiple attributes in one index, in the example there is only a single for each, but you can have an array in general. I.e. if there were two different 3D volumes of silicon, there could be a single index referring to the combined volume of both regions.

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,542 @@
+boundary conditions. *Palace* offers numerous boundary condition types, all
+documented in the [boundaries reference](config/boundaries.md).
+
+Unlike 3D regions which all require `Material` specifications, 2D surfaces have
+a default behavior: any surface without an explicit boundary condition is
+treated as a Perfect Magnetic Conductor (PMC), where the tangential component of
+the magnetic field is zero.
```
This isn't exactly correct, any *external* boundary with no condition set is marked as PMC (whatever the natural boundary condition is strictly speaking (`n . grad phi = 0` or `n x curl E = 0` or `n x curl A = 0`)), an internal boundary has no condition applied. This is because the terms from left and right will cancel on such a boundary, so not explictly adding something is equivalent to doing nothing.

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,542 @@
+  - Two `Terminal` surfaces (one for each sphere)
+
+`Terminals` are particularly important for `Electrostatic` simulations. *Palace*
+activates each terminal sequentially, applying a unit of potential and solving
```
, when a terminal is not activated, it is held at ground (V = 0).

hughcars commented on docs/src/quick.md:
````diff
@@ -0,0 +1,542 @@
+Once the simulation is completed, you'll find a `postpro` directory containing:
+
+```@example json
+logrun(`tree postpro`); # hide
````
I get an error for this section after having all the previous (including the run) work
```
Base.IOError("could not spawn `tree postpro`: no such file or directory (ENOENT)", -2)
```
I assume this is because I ran with `cd docs && julia ` then `] activate .` followed by `include("make.jl)`. I presume this can be done more neatly to pull this data correctly.

Sbozzolo replied:
Ah! I found out that macOS does not come with `tree` apparently. I'll get this in a different way

Sbozzolo replied:
I changed to just use `ls`

hughcars reviewed (changes requested):
Generally looks pretty good to me. Some minor wordsmithing and one possible issue with generating the html output of the postpro tree view, but basically ready.

Sbozzolo commented:
> Generally looks pretty good to me. Some minor wordsmithing and one possible issue with generating the html output of the postpro tree view, but basically ready.

I rebased against main, and addressed the comments.

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,548 @@
+
+In this tutorial, we will:
+
+ 1. [Install *Palace* using Spack](#installing-palace)
```
Have you tried these links? When I click I get `file:///Users/hughcars/AWS/palace/docs/build/quick/index.html#installing-palace` (on my local build, from chrome) it doesn't actually move to the correct section below. The links are case sensitive despite what the vscode markdown plug in says `#Installing-Palace` works. 

This applies to all the markdown links I believe.

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,548 @@
+equation for electric potential sequentially activating all the `Terminals` on
+the mesh while setting the non-activated terminals to ground. All simulation types
+in *Palace* have some form of iteration (over frequencies, times, mode numbers,
+or terminals). We save the output in the `postpro` folder, which will be
```
Save in the `"Output"` folder from the `"Problem"` section, in this example `"postpro"`. Basically want to make sure this is an obviously configurable thing

hughcars commented on docs/src/quick.md:
```diff
@@ -0,0 +1,548 @@
+in *Palace* have some form of iteration (over frequencies, times, mode numbers,
+or terminals). We save the output in the `postpro` folder, which will be
+overwritten if it already exists. See the [problem types
+documentation](config/problem.md#problem-types) for details on all available
```
In the rest of the documentation we are linking to pieces within the configuration file with their nested format
`config["Boundaries"]["PEC"]["Attributes"]` for example, to highlight in each link the location within the containing tree. I think most of the links to sections of the `config/` directory could be upgraded to links of that form (and also checked, I think most of these links aren't working due to the case sensitivity).

hughcars reviewed (changes requested):
Took the chance to read through again and found a few more small issues: most links are broken and there's a few minor grammar issues. Additionally the `spheres.json` file shouldn't be modified, as we don't want to encourage reducing the verbosity and eliding material properties.
